runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1277,7 +1280,13 @@ private void onNewEventTimer(TimerData newTimer) {
           "Timer with id %s is not an event time timer!",
           newTimer.getTimerId());
       if (timerUsesOutputTimestamp(newTimer)) {
-        outputTimestampQueue.add(newTimer.getOutputTimestamp().getMillis());
+        outputTimestamps.compute(
+            newTimer.getOutputTimestamp().getMillis(),
+            (k, v) -> {
+              Set<String> timerIds = v == null ? new HashSet<>() : v;
+              timerIds.add(getContextTimerId(newTimer.getTimerId(), newTimer.getNamespace()));
```

Comment:
IMHO there is no need to store the timer id here.

Reply:
It should not. But in some cases, a timer output timestamp seems to appear in the queue _twice_ (for the same timer) and removed only _once_, which causes the pipeline to get stuck.

Reply:
Duplicates are ok. If we have multiple timers set with different ids or even for the same keys, duplicates are expected. We could use a TreeMap instead of a PriorityQueue like we do in FlinkTimerInterals for the WatermarkHoldState, but I don't see anything wrong with the implementation here. 

You likely have a problem in your application code, i.e. you set a timer output timestamp which holds back the watermark but your fire timestamp has not been reached by the watermark yet. Please investigate that first. Also note #12531 which was a bug in the Python SDK.